test/core/security/credentials_test.cc

```diff
@@ -449,126 +449,141 @@ static void test_oauth2_token_fetcher_creds_parsing_missing_token_lifetime(
-                                  &error)) {
-    // Synchronous result.  Invoke the callback directly.
-    check_request_metadata(state, error);
-    GRPC_ERROR_UNREF(error);
+
+  void CheckRequestMetadata(grpc_error_handle error) {
+    gpr_log(GPR_INFO, "expected_error: %s",
+            grpc_error_std_string(expected_error_).c_str());
+    gpr_log(GPR_INFO, "actual_error: %s", grpc_error_std_string(error).c_str());
+    if (expected_error_ == GRPC_ERROR_NONE) {
+      GPR_ASSERT(error == GRPC_ERROR_NONE);
+    } else {
+      std::string expected_error;
+      GPR_ASSERT(grpc_error_get_str(expected_error_, GRPC_ERROR_STR_DESCRIPTION,
+                                    &expected_error));
+      std::string actual_error;
+      GPR_ASSERT(
+          grpc_error_get_str(error, GRPC_ERROR_STR_DESCRIPTION, &actual_error));
+      GPR_ASSERT(expected_error == actual_error);
+      GRPC_ERROR_UNREF(expected_error_);
+    }
+    gpr_log(GPR_INFO, "expected_size=%" PRIdPTR " actual_size=%" PRIdPTR,
+            expected_.size(), md_array_.size);
+    GPR_ASSERT(md_array_.size == expected_.size());
+    CheckMetadata(expected_, &md_array_);
+    delete this;
   }
-}
+
+  static void CheckMetadata(const std::map<std::string, std::string>& expected,
+                            grpc_credentials_mdelem_array* md_array) {
+    for (auto const& i : expected) {
+      size_t j;
+      for (j = 0; j < md_array->size; ++j) {
+        absl::string_view actual_key =
+            grpc_core::StringViewFromSlice(GRPC_MDKEY(md_array->md[j]));
+        if (actual_key == i.first) {
+          absl::string_view actual_value =
+              grpc_core::StringViewFromSlice(GRPC_MDVALUE(md_array->md[j]));
+          GPR_ASSERT(actual_value == i.second);
+          break;
+        }
+      }
+      if (j == md_array->size) {
+        gpr_log(GPR_ERROR, "key %s not found", i.first.c_str());
+        GPR_ASSERT(0);
+      }
+    }
+  }
+
+ private:
+  grpc_error_handle expected_error_;
+  std::map<std::string, std::string> expected_;
+  grpc_credentials_mdelem_array md_array_;
+  grpc_closure on_request_metadata_;
+  grpc_polling_entity pollent_;
+};
+
+}  // namespace
 
 static void test_google_iam_creds(void) {
   grpc_core::ExecCtx exec_ctx;
-  expected_md emd[] = {{GRPC_IAM_AUTHORIZATION_TOKEN_METADATA_KEY,
-                        test_google_iam_authorization_token},
-                       {GRPC_IAM_AUTHORITY_SELECTOR_METADATA_KEY,
-                        test_google_iam_authority_selector}};
-  request_metadata_state* state =
-      make_request_metadata_state(GRPC_ERROR_NONE, emd, GPR_ARRAY_SIZE(emd));
+  std::map<std::string, std::string> emd = {
+      {GRPC_IAM_AUTHORIZATION_TOKEN_METADATA_KEY,
+       test_google_iam_authorization_token},
+      {GRPC_IAM_AUTHORITY_SELECTOR_METADATA_KEY,
+       test_google_iam_authority_selector}};
+  RequestMetadataState* state =
+      RequestMetadataState::NewInstance(GRPC_ERROR_NONE, emd);
```

Comment:
`std::move()` to pass `emd` here.  Same for all tests.

Alternatively, change `RequestMetadataState` to just take a pointer to the map.